### ISSUE TITLE
QUDA CPS interface support for clover fermions

### ISSUE THREAD

**Issue by maddyscientist** [lattice/quda#120]
The current QUDA CPS interface supports only Wilson or domain-wall fermions, with no support for clover fermions.  This is use case required for RIKEN to deploy their current computations.

While not really a QUDA issue, I include this here to track this feature's inclusion into CPS.

**Comment by cpviolator** [comment]
Mike,

Please see the REPO published here https://github.com/cpviolator/QUDA-CPS For a version of CPS that has QUDA support for Clover fermions.

**Comment by mikeaclark** [comment]
Thanks @cpviolator!  This is a nice surprise.  Can you tell me:
* Does this support both asymmetric and symmetric clover fermions?
* Have you offloaded the clover generation to QUDA and / or the clover inversion to QUDA?
* The README mentions that Wilson is only supported through setting the clover coefficient to 0.0.  So is this a different code path from the Wilson support I added to CPS years (and years) ago?
* Any changes needed or requested for QUDA to help this?
* Are there plans to get this into mainline CPS?

**Comment by cpviolator** [comment]
@mikeaclark,

My pleasure. I'll answer as best I can.

1. We've never run on anything but symmetric lattices with this code. However...
2. We take the clover data from CPS and make the necessary conversions in the functions 'fill_clover_all' and 'quda_clover_interface', so the clover generation is done entirely in CPS and inserted to QUDA via loadCloverQuda(). We take the same route with the gauge field. We found that we had to do some reordering of the SU(3) matrices to get things working. All this is done in CPS via 'gauge_reorder()' and loaded with loadGaugeQuda(). The fermion fields (due to checkerboading) also need reordering, again, done in CPS. All that is left is a call to invertQuda(X, B, params). I will tentatively assume that asymmetric lattices will be supported this way, but I make no guarantee.
3. I see that all Wilson (and DWF) support is done on QUDA. I'm in the middle of hooking up Wilson and DWF to QUDA via the existing routines. We took the route we did because...
4. QUDA, as I'm sure you are very much aware, takes an eon to compile, which is why working out the clover support on CPS first and leaving QUDA untouched was the better route. I would be very happy to try to port the routines into QUDA in the spirit of open source. I could use a little help sidestepping CPS's archaic RR eigensolver routines and hooking up with QUDA's Lanczos though!
5. I have no idea about folding this into mainline CPS. To be honest, I think the better route would be to fold these routines into QUDA and limit the CPS clover interface to parameter settings and a function call.

**Comment by cpviolator** [comment]
@mikeaclark,

There is one problem I'm having with regards to QUDA 0.7.x. When we compile QUDA-CPS against 0.7.x, QUDA throws an error:
Error: param tol not defined (check_params.h:145 CheckInvertParams())
but we do not see this problem when using 0.5.x or 0.6.x. This is confusing because the QUDA parameter 'tol' is always defined and and line 145 in check_params.h has nothing to do with the tol parameter.

**Comment by mikeaclark** [comment]
1. Sorry I wasn't clear on symmetric / asymmetric. I meant with respect to preconditioning, e.g., from enum_quda.h
```c++
  typedef enum QudaMatPCType_s {
    QUDA_MATPC_EVEN_EVEN,
    QUDA_MATPC_ODD_ODD,
    QUDA_MATPC_EVEN_EVEN_ASYMMETRIC,
    QUDA_MATPC_ODD_ODD_ASYMMETRIC,
    QUDA_MATPC_INVALID = QUDA_INVALID_ENUM
  } QudaMatPCType;
```
2. For the gauge field, doesn't setting the gauge order to `QUDA_CPS_GAUGE_ORDER` work for you?  This works fine for DWF / Wilson, as far I'm aware.  For the fermion field, doesn't `QUDA_CPS_WILSON_DIRAC_ORDER` work?  If you pass null pointers to the `loadCloverQuda()` interface then it will construct the clover field and do the inverse in QUDA on the GPU, so avoiding CPU overhead and removing all the CPU memory traffic.  It doesn't work yet with anisotropic lattices, like Jlab have in the time, but for isotropic lattice it should just work.
3. For advice on QUDA's eigen-vector routines, I refer you to @alexstrel :)
4. You should find for compilation time is getting better with the most recent releases of QUDA.  In particular I would advise you at present to work from the `develop` branch since this is considered stable as this has some compilation time improvements not in `release/0.7.x`.  Also when building, make sure you build in parallel with `make -j N` where `N` is the number of hardware threads on your CPU (assuming you have enough memory).
5. If any additional reordering support is needed then I'm happy to incorporate it.

I can add you to the commit list on github.  We're always happy to have more developers.  Our development model is such that master is always the latest official release, hotfixes for the next point releases are in `release/x.y.z` with new features being incorporated into the `develop` branch which is kept stable.  All merges into `master`, `release/x.y.z` and `develop` are done via pull request, review and merge.  Please join the fun!

On your `tol` issue, can you confirm that this is fresh build, and you don't have any stale `quda.h` incorporated into a CPS object file?  As always with QUDA, if there's an issue, please file an issue.

**Comment by cpviolator** [comment]
* I'll do some testing with the `_MATPC_` params and see what needs to be done on the CPS end to include them.
* The Quda `gauge_order` parameter we pass to QUDA already `QUDA_CPS_GAUGE_ORDER` so I will certainly try that change out, and again we use `QUDA_DIRAC_ORDER` for the `dirac_order` so I will experiment there too and update accordingly. We can scratch together some small anisotropic lattices on the cpu with clover on CPS and test results our interface code with GPU. A small project that might be beneficial.
* I will contact alexstrel about Lanczos, thanks for putting me in the right direction!
* I'll switch to the `develop` branch, make everything from fresh from repo and retest. If the `tol` issue persists, I'll submit an issue with more details.

Thanks for your help @mikeaclark! I hope we can get clover support in time for 0.8.0!

**Comment by mikeaclark** [comment]
You are welcome.  :+1: 

Just so I fully understand your comment, what do you mean by clover support for 0.8.0?  What is needed on the QUDA end that isn't already there?  

r are you referring to anisotropic support for clover field construction since it only support isotropic at the moment.  The clover inversion should work already for anisotropic (I think).

**Comment by cpviolator** [comment]
Oops! I just meant closing issue #120 before the 0.8.0 release. It's been a long day! Sorry!

I'll get on with what's been discussed here and update the thread when I have results. Thanks again!

**Comment by maddyscientist** [comment]
@cpviolator Just checking on the status of CPS and clover.  Is there anything that needs to be done at the QUDA end?

**Comment by cpviolator** [comment]
Not to my knowledge. I'm running Clover fermions on CPS with QUDA support
with no problems and no bottlenecking, using the
QUDA_CPS_WILSON_DIRAC/GAUGE_ORDER flags and a little vector rearranging on
the host. If I find a more elegant method I'll make a branch then pull
request.

**Comment by maddyscientist** [comment]
Thanks.

Did you ever try the built in clover construction / inversion to avoid the need to do any clover transfers?  Also, we could support the reordering as part of QUDA directly to keep CPS cleaner.

Is this in mainline CPS?

**Comment by cpviolator** [comment]
I'm hacking the public release, I'm not affiliated with Columbia.

I will make a priority of streamlining the clover construction / inversion this week. It's my understanding that passing null pointers in loadCloverQuda will perform the construction / inversion on the device, a nice feature that ought to be utilised.

**Comment by AlexVaq** [comment]
Yes, null pointers in loadCloverQuda forces the gpu to calculate the Fmunu tensor and the clover term.

**Comment by maddyscientist** [comment]
Great.  Probably best to check it in two steps: first just pass a null pointer for the inverse to check the inversion works ok, then check the full clover construction.  It should just work: only reason it wouldn't work is if you're using anisotropic fields, which we don't support yet (though this would only affect the clover construction, not the inversion which should work regardless).  Something I need to do is get this support added.

Can you test this using the release/0.8.x branch, as that's the one we're cleaning up at the moment and bug fixing prior to testing QUDA 0.8.  Thanks.

**Comment by cpviolator** [comment]
@maddyscientist will do re 0.8.x.

**Comment by mathiaswagner** [comment]
Any updates here? Any fixes required for 0.8.0 ?

**Comment by cpviolator** [comment]
At the moment, I can pass the gauge field and the fermion fields to QUDA with the relevant QUDA settings, but I must still construct the clover terms on the CPU then load. Passing null pointers to `loadCloverQuda`,

`loadCloverQuda(null_pointer, null_pointer, &quda_inv_params)`

in an attempt to have the device construct the clover matrices results in the solver diverging on the first iteration. However, the device clover inversion routine works perfectly, i.e.,

`loadCloverQuda(my_clover_matrices, null_pointer, &quda_inv_params)`

gives correct results. My QUDA version is 0.7.1.

**Comment by maddyscientist** [comment]
Ok thanks for the update @cpviolator.  I will take a look at this when I get a chance.

Is it any faster in your workflow to invert on the GPU rather than transfer the inverse?

**Comment by cpviolator** [comment]
I will time it now and post back with results.

**Comment by maddyscientist** [comment]
Also, do you think the test you ran with CPS is something I can easily reproduce?  I see you've posted your variant of CPS on github...

**Comment by cpviolator** [comment]
Let me update it properly before you download a copy. Reproduction will be straightforward with some simple 1 hadron simulation code.

**Comment by cpviolator** [comment]
Ok, I've updated the QUDA-CPS repo. The current version will construct the clover matrices on the host, then have the device calculate the clover inverse matrices. You can change this by editing the function `loadCloverQuda(ptr1,ptr2,&params)`
in the file
`QUDA-CPS_v5_0_26/src/util/quda_interface/quda_interface.C`
The single hadron code should compile, but packaging that was handed off a project collaborator. If you have problems with the hadron code, let me know and I'll send you a much simpler set of files for testing.

**Comment by cpviolator** [comment]
Here is some simple 1pion code. Just adjust the makefile so that it knows
where your QUDA, CUDA, and CPS builds are, and also comment out #define READ in
the 1pion.cpp and 1pion.GPU.cpp files on your first run. This will generate
a small lattice that you can read in on subsequent test runs.

**Comment by cpviolator** [comment]
In the repo I added a directory called `1pion_quick_test` that should be a lot easier work with.

**Comment by maddyscientist** [comment]
Thanks.  Not sure when I'll get to this, but hopefully soon.  I shortly have to turn my attention to clover multigrid once I've finished my current work on applying even/odd preconditioning to the multigrid framework, so that would probably involve minimal context switching.

**Comment by maddyscientist** [comment]
A couple of quick questions in the meantime though:
* are you doing asymmetric or symmetric preconditioning?
* even-even or odd-odd?
* the lattice is isotropic I presume?
* are you using gauge reconstruction?

**Comment by cpviolator** [comment]
Information from CPS on preconditioning:
Algorithm:
i.  The preconditioned matrix is inverted with
i.1: on the odd checkerboard:  the conjugate gradient 
i.2: on the even checkerboard: explicit inversion of those local and hermitian clover matrices.
ii. steps: (no extra storage required)
1. out_even0 = Aee^inv in_even
2. out_odd  = (MatPcDagMatPc)^inv MatPcDag 
                                (kappa Doe out_even0 + in_odd)
3. out_even  = Aee^inv (kappa Deo out_odd + in_even)`

* The lattices I use are indeed isotropic, but Anisotropic lattices are looming on the horizon...
* The gauge reconstruction is 12.

**Comment by maddyscientist** [comment]
Thanks for the info.  So when solving, are you passing `in` to QUDA and using say `QUDA_MAT_SOLUTION` with QUDA_DIRECT_PC_SOLVE to compute the full `out`, or are you just letting QUDA do the preconditioned solve, `QUDA_MATPC_SOLUTION`, i.e., just passing to QUDA `in_odd` and getting back `out_odd`?

**Comment by cpviolator** [comment]
I pass the full `in` and 'out` vectors to `invertQuda(out,in,&inv_param) with the QUDA flags:

`inv_param.solution_type = QUDA_MAT_SOLUTION;`
`inv_param.solve_type = QUDA_NORMOP_PC_SOLVE;`
`inv_param.matpc_type = QUDA_MATPC_EVEN_EVEN;`

So QUDA does all the work.

**Comment by maddyscientist** [comment]
Ok.  I'll take a look, thanks for all the info.

If you have a chance, I've got one final test for you do.  Can you retest but with `QUDA_RECONSTRUCT_NO`, just to rule out reconstruction as a possible source of error.

One thing, it looks like the type of preconditioning you're requesting there is not the same as the CPS is doing.  `QUDA_EVEN_EVEN` / `QUDA_ODD_ODD` will invert both parities clover matrix, where as `QUDA_EVEN_EVEN_ASYMMETRIC` / QUDA_ODD_ODD_ASYMMETRIC` will do what the CPS is doing.  Not that this matters, since you're requesting the full solution, which is obviously the same regardless of the underlying preconditioner.

**Comment by cpviolator** [comment]
Adjusting the `QUDA_RECONSTRUCT_XX` flag to NO has the same effect and the solver diverges.

Thanks for the heads up about preconditioning. When I time the clover inverse routines I'll do some timing tests to see what the differences are with respect to preconditioning.

**Comment by maddyscientist** [comment]
Just looking at your interface, and the first thing I've spotted is that you're not enabling the the autotuning.  This will greatly reduce the performance of the linear solver.
```
  inv_param.tune = QUDA_TUNE_NO;
```
Any reason you're not using mixed precision or BiCGstab?  There's probably a factor of 3-4 sitting untapped there.

**Comment by cpviolator** [comment]
This is on my list of jobs before scaling up to large lattices. Indeed, Joel (my supervisor) has strongly suggested I move to Mixed ASAP. I've taken the policy of trying to keeping everything as accurate as possible, but once we move to L=16,20,24... then yes, mixed precision will be employed. I didn't know about the `inv_param.tune = QUDA_TUNE_NO;` flag, it's probably been that way since Joel and I wrote the interface.

Tuning and benching this code is obviously very important, so your help here will be invaluable, thank you. I'll include Mixed Vs Double prec. in the timing study.

**Comment by maddyscientist** [comment]
Using mixed precision won't affect the final result at all, only make it go faster.  It's a violation of the no-free-lunch theorem :wink: 

For double-single, using `reliable_delta = 0.01` should work fine, and for double-half, `reliable_delta = 0.1` is probably optimal.  Both should give a big speedup, with double-half precision working fastest.

**Comment by cpviolator** [comment]
That's awesome, thank you! Given all this new input, I'll bench the code properly and write a short report. I have to do it for thesis anyway so I may as well do a proper job. Tabulated results will appear here for:

*double-double / double-single / double-half,
*comparisons with `QUDA_TUNE_NO/YES`,
*timings for host Vs CPU clover matrix inversion,

all done at a variety of L and m_{\pi} in the quenched approximation. Data for dynamical lattices too, but only for those lattices I have.

BTW, I read in your [conference paper](http://arxiv.org/pdf/1011.2775.pdf) that you used the Hadron Spectrum Collaboration's Ainsotropic Clover lattices for your testing. Are these lattices available on FNAL? How would one go about getting hold of them?

**Comment by maddyscientist** [comment]
I don't know if the entire set is available or not, @bjoo can you help out here?

**Comment by maddyscientist** [comment]
Also, when running with tuning, for a fair comparison, you need to set the `QUDA_RESOURCE_PATH` environment variable and compare multiple runs.  See [here](https://github.com/lattice/quda/wiki/QUDA-Quick-Start-Guide#kernel-autotuning) for more details.

**Comment by mathiaswagner** [comment]
OT: Just realized that the tuning is describe twice on that wiki page. (above internal testing and in the corresponding section). Anyhow, we'll rewrite a lot here for cmake ...

**Comment by cpviolator** [comment]
May I ask, how does one set that? I put
`export QUDA_RESOURCE_PATH=/home/howarth/QUDA_resources` 
in my .bash_profile and started a new shell, but QUDA doesn't recognise it. Am I doing something terribly wrong?

**Comment by maddyscientist** [comment]
Can you just set it in the shell directly and check that it's being correctly:
```
export QUDA_RESOURCE_PATH=...
$PATH_TO_QUDA/tests/dslash_test # should run tuning and save out cache to tunecache.tsv
$PATH_TO_QUDA/tests/dslash_test # should load generated tunecache.tsv
```

**Comment by mathiaswagner** [comment]
check with 'env' that it is correctly set in your shell.

**Comment by cpviolator** [comment]
Thank you, `QUDA_RESOURCE_PATH` is now being set and recognised. And it looks like double-half using CG gives a factor two improvement! I see there is a whole host of QUDA inverters so I'll have a lot to play with with regards to benching the code.

**Comment by maddyscientist** [comment]
Excellent.  Generally, BiCGstab gives the best performance for clover, and I would expect a further factor of two improvement.  Until multigrid is stable anyway...

**Comment by mathiaswagner** [comment]
While you are benchmarking anyway you can also try to set different reconstruct options for sloppy precision (half, single) and full precision (double). At least for staggered I sometimes saw better performance by using no_reconstruct for full precision and reconstruct 12 / 8 for sloppy (technically 13 / 9 for staggered). No huge effects but while you are benchmarking anyway it might be nice to get some more data points for this.

**Comment by cpviolator** [comment]
Thanks @mathiaswagner, the more angles I have to look at, the better!

I will also be looking at swapping out the CPS clover (and Wilson) inverter in the HMC phi algorithm. Right now I use a BG/Q to generate dynamical configurations, but moving that over to the GPU would be great. It would make for a solid variant of CPS that can generate publication quality data on just one or two GPU cards.

**Comment by mathiaswagner** [comment]
@cpviolator @maddyscientist PING: Any updates on this? Something we need to fix for 0.8?

**Comment by maddyscientist** [comment]
Closing this ancient bug.  Clover will work with CPS already, using the regular Wilson interface will suffice.  With modern QUDA there is no need to support the separate clover interface since we can just compute the clover field on the GPU and this is faster than downloading it.
